### PR TITLE
Add FP16 Support for SageMaker Model Parallel

### PR DESCRIPTION
**Pull request by haohanchen-aws**
(This PR is still pending some changes/fixes)
This PR adds support for SageMaker Model Parallel with FP16.

- To Enable fp16 with SMMP, user needs to add `fp16: True` to `SM_HP_MP_PARAMETERS`, when there's mismatch between `SM_HP_MP_PARAMETERS` and trainer args, a warning log would be printed and `SM_HP_MP_PARAMETERS` will be used as truth.
- Remove amp related stuff for SMMP as SMMP manages it's own half precision.
- Added `clip_master_grads` for grad clipping
- Some minor fixes

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by rahul003** [comment]
@philschmid Please review

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -506,14 +520,14 @@ def __init__(
             logger.info(f"Using {args.half_precision_backend} half precision backend")
 
         self.do_grad_scaling = False
-        if (args.fp16 or args.bf16) and not args.deepspeed:  # deepspeed manages its own half precision
+        if (args.fp16 or args.bf16) and not (
+            args.deepspeed or is_sagemaker_mp_enabled()
+        ):  # deepspeed and SageMaker Model Parallel manage their own half precision
```
Comment (sgugger) [inline src/transformers/trainer.py:525]:
```suggestion
        if (args.fp16 or args.bf16) and not (args.deepspeed or is_sagemaker_mp_enabled()):
            # deepspeed and SageMaker Model Parallel manage their own half precision  
```
Better like this.

**Review by sgugger — APPROVED** [review]
This simplifies things quite a lot, thanks! Just have one small styling comment.

File: src/transformers/trainer.py (hunk 1 above)
Comment (haohanchen-yagao) [reply src/transformers/trainer.py]: Updated with the reformatting suggestion, thanks!

**Comment by sgugger** [comment]
Can you just run `make style` on your branch to fix the code quality issue? Thanks!

**Pull request merged**